lib/Ast.ml

```diff
@@ -1,223 +0,0 @@
-(** Abstract syntax. *)
-open Core_kernel
-
-(* for auto generating s-exp *)
-
-(* == Source code locations for semantic errors == *)
-type location =
-  | Location of Lexing.position * Lexing.position  (** delimited location *)
-  | Nowhere  (** no location *)
-
-(* == Unsized types == *)
-type originblock =
-  | Primitives
-  | Functions
-  | Data
-  | TData
-  | Param
-  | TParam
-  | Model
-  | GQuant
-
-and unsizedtype =
-  | Int
-  | Real
-  | Vector
-  | RowVector
-  | Matrix
-  | Array of unsizedtype
-  | Fun of (originblock * unsizedtype) list * returntype
-  | PrimitiveFunction
-
-and returntype = Void | ReturnType of unsizedtype
-
-(* == Expressions == *)
-and identifier = {name: string; id_loc: location sexp_opaque [@compare.ignore]}
-
-and infixop =
-  | Plus
-  | Minus
-  | Times
-  | Divide
-  | Modulo
-  | LDivide
-  | EltTimes
-  | EltDivide
-  | Exp
-  | Or
-  | And
-  | Equals
-  | NEquals
-  | Less
-  | Leq
-  | Greater
-  | Geq
-
-and prefixop = Not | UMinus | UPlus
-
-and postfixop = Transpose
-
-and 'e index =
-  | All
-  | Single of 'e
-  | Upfrom of 'e
-  | Downfrom of 'e
-  | Between of 'e * 'e
-  | Multiple of 'e
-
-and 'e expression =
-  | Conditional of 'e * 'e * 'e
-  | InfixOp of 'e * infixop * 'e
-  | PrefixOp of prefixop * 'e
-  | PostfixOp of 'e * postfixop
-  | Variable of identifier
-  | IntNumeral of string
-  | RealNumeral of string
-  | FunApp of identifier * 'e list
-  | CondFunApp of identifier * 'e list
-  | GetLP
-  (* deprecated *)
-  | GetTarget
-  | ArrayExpr of 'e list
-  | RowVectorExpr of 'e list
-  | Paren of 'e
-  | Indexed of 'e * 'e index list
-
-and expression_untyped_metadata =
-  {expr_untyped_meta_loc: location sexp_opaque [@compare.ignore]}
-
-and expression_typed_metadata =
-  { expr_typed_meta_origin_type: originblock * unsizedtype
-  ; expr_typed_meta_loc: location sexp_opaque [@compare.ignore] }
-
-and untyped_expression =
-  | UntypedExpr of (untyped_expression expression * expression_untyped_metadata)
-
-and typed_expression =
-  | TypedExpr of (typed_expression expression * expression_typed_metadata)
-
-(* == Statements == *)
-and assignmentoperator =
```

Comment:
Assignment operators are basically the same as the infix operators (minus the arrow) - so we could probably factor this such that an assignment takes an infix operator (where conditional operators are separated out).

Reply:
Done.